lib/srv/alpnproxy/kube.go

```diff
@@ -66,13 +98,94 @@ func (m *KubeMiddleware) OverwriteClientCerts(req *http.Request) ([]tls.Certific
 		return nil, trace.BadParameter("expect a TLS request")
 	}
 
+	m.certsMu.RLock()
 	cert, ok := m.certs[req.TLS.ServerName]
+	m.certsMu.RUnlock()
 	if !ok {
 		return nil, trace.NotFound("no client cert found for %v", req.TLS.ServerName)
 	}
+
+	newCert, err := m.ensureCertValidity(req.Context(), cert, req.TLS.ServerName)
+	if err != nil {
+		// If we exceeded timeout, we return an error that will try to get user attention to the local proxy
+		if errors.Is(err, ErrUserInputRequired) {
+			return nil, trace.Wrap(&apierrors.StatusError{
+				ErrStatus: metav1.Status{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Status",
+						APIVersion: "v1",
+					},
+					Status:  metav1.StatusFailure,
+					Code:    http.StatusGatewayTimeout,
+					Reason:  metav1.StatusReasonTimeout,
+					Message: "Local Kube proxy requires user input to continue",
+				},
+			})
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	// If certificate was reissued we use it to process this request.
+	if newCert != nil {
+		cert = *newCert
+	}
+
 	return []tls.Certificate{cert}, nil
 }
 
+var ErrUserInputRequired = errors.New("user input required")
+
+func (m *KubeMiddleware) ensureCertValidity(ctx context.Context, cert tls.Certificate, serverName string) (*tls.Certificate, error) {
```

Comment:
godoc and rename the function as it reissues a new certificate if it's not valid